File: Battleship-game.py
import os

MIN_BOARD_SIZE = 5
MAX_BOARD_SIZE = 15
MAX_SHIP_TYPES = 5
MAX_NAME_LENGTH = 50

class Ship:
    def __init__(self, size):
        self.size = size
        self.hits = 0
        self.isSunk = False

class Player:
    def __init__(self, name, ship_count):
        self.name = name
        self.score = 0
        self.ships = [Ship(i + 3) for i in range(ship_count)]

def create_board(size):
    return [['~' for _ in range(size)] for _ in range(size)]

def free_board(board):
    pass  # Python'da garbage collector kullanıldığı için manuel bellek temizleme gerekmez

def initialize_boards(board1, board2):
    for i in range(len(board1)):
        for j in range(len(board1[0])):
            board1[i][j] = '~'
            board2[i][j] = '~'

def print_hidden_board(board):
    size = len(board)
    print("   " + " ".join(str(i + 1) for i in range(size)))
    
    for i in range(size):
        print(chr(ord('A') + i) + " |", end="")
        for j in range(size):
            if board[i][j] == 'X' or board[i][j] == 'O':
                print(board[i][j], end=" ")
            else:
                print("~", end=" ")
        print()

def print_player_info(player, player_name):
    print(f"{player_name}'s Info:")
    print(f"Score: {player.score}")
    for i, ship in enumerate(player.ships):
        print(f"Ship {i + 1} - Size: {ship.size}, Hits: {ship.hits}")

def place_ship(player, board, size, ship_index):
    while True:
        try:
            coords = input(f"Enter the starting coordinates for Ship {ship_index + 1} (size {player.ships[ship_index].size}): ").split()
            row, y = coords[0].upper(), int(coords[1])
            x = ord(row) - ord('A')

            direction = int(input("Downward (1) or Rightward (2)? "))

            if direction == 1:
                if x + player.ships[ship_index].size > size:
                    print("Invalid placement. Try again.")
                else:
                    for i in range(player.ships[ship_index].size):
                        if board[x + i][y - 1] != '~':
                            print("Invalid placement. Try again.")
                            break
                        else:
                            board[x + i][y - 1] = str(ship_index + 1)
                    break
            elif direction == 2:
                if y + player.ships[ship_index].size > size:
                    print("Invalid placement. Try again.")
                else:
                    for i in range(player.ships[ship_index].size):
                        if board[x][y + i - 1] != '~':
                            print("Invalid placement. Try again.")
                            break
                        else:
                            board[x][y + i - 1] = str(ship_index + 1)
                    break
            else:
                print("Invalid direction. Try again.")
        except (ValueError, IndexError):
            print("Invalid input. Try again.")

def place_ships(player, board, size, ship_count):
    for i in range(ship_count):
        player.ships[i].size = i + 3
        player.ships[i].hits = 0
        player.ships[i].isSunk = False

        print_hidden_board(board)
        print_player_info(player, "Player")
        place_ship(player, board, size, i)
        os.system("clear")  # Ekranı temizle

def make_move(player, opponent_board, size, ship_count):
    while True:
        try:
            coords = input("Enter the coordinates to make a move: ").split()
            row, y = coords[0].upper(), int(coords[1])
            x = ord(row) - ord('A')

            if not (0 <= x < size and 1 <= y <= size):
                print("Invalid coordinates. Try again.")
                continue

            if opponent_board[x][y - 1] == 'X' or opponent_board[x][y - 1] == 'O':
                print("Already guessed that location. Try again.")
                continue

            if opponent_board[x][y - 1] != '~':
                ship_index = int(opponent_board[x][y - 1]) - 1
                print("Hit!")
                opponent_board[x][y - 1] = 'X'
                player.ships[ship_index].hits += 1

                if player.ships[ship_index].hits == player.ships[ship_index].size:
                    print("You sank a ship!")
                    player.ships[ship_index].isSunk = True
                    player.score += player.ships[ship_index].size

                    all_sunk = all(ship.isSunk for ship in player.ships)
                    if all_sunk:
                        print("All opponent's ships are sunk! You win!")
                        return False  # Oyunu bitir
            else:
                print("Miss!")
                opponent_board[x][y - 1] = 'O'

            return True  # Oyun devam ediyor
        except (ValueError, IndexError):
            print("Invalid input. Try again.")

def main():
    board_size = int(input(f"Enter the board size (between {MIN_BOARD_SIZE} and {MAX_BOARD_SIZE}): "))

    if not MIN_BOARD_SIZE <= board_size <= MAX_BOARD_SIZE:
        print("Invalid board size. Exiting...")
        return 1

    ship_count = int(input("Enter the number of ships: "))

    if not 1 <= ship_count <= MAX_SHIP_TYPES:
        print("Invalid number of ships. Exiting...")
        return 1

    player1_board = create_board(board_size)
    player2_board = create_board(board_size)

    player1 = Player(input("Enter Player 1's name: "), ship_count)
    player2 = Player(input("Enter Player 2's name: "), ship_count)

    print(f"{player1.name}, place your ships:")
    place_ships(player1, player1_board, board_size, ship_count)

    print(f"{player2.name}, place your ships:")
    place_ships(player2, player2_board, board_size, ship_count)

    current_player = 1

    while True:
        if current_player == 1:
            os.system("clear")
            print(f"{player1.name}'s Turn:")
            print_hidden_board(player2_board)
            print_player_info(player1, player1.name)
            if not make_move(player1, player2_board, board_size, ship_count):
                break  # Oyun bitti
        else:
            os.system("clear")
            print(f"{player2.name}'s Turn:")
            print_hidden_board(player1_board)
            print_player_info(player2, player2.name)
            if not make_move(player2, player1_board, board_size, ship_count):
                break  # Oyun bitti

        current_player = 3 - current_player  # Sıradaki oyuncu

    print("Final Scores:")
    print_player_info(player1, player1.name)
    print_player_info(player2, player2.name)

    free_board(player1_board)
    free_board(player2_board)

if __name__ == "__main__":
    main()

File: Battleship-game.c
#include <stdio.h>
#include <stdlib.h>

#ifdef _WIN32
#define CLEAR_SCREEN system("cls")
#else
#define CLEAR_SCREEN system("clear")
#endif

#define MIN_BOARD_SIZE 5
#define MAX_BOARD_SIZE 15
#define MAX_SHIP_TYPES 5
#define MAX_NAME_LENGTH 50

typedef struct {
    int size;
    int hits;
    int isSunk;
} Ship;

typedef struct {
    char name[MAX_NAME_LENGTH];
    int score;
    Ship ships[MAX_SHIP_TYPES];
} Player;

char **createBoard(int size) {
    char **board = (char **)malloc(size * sizeof(char *));
    for (int i = 0; i < size; ++i) {
        board[i] = (char *)malloc(size * sizeof(char));
        for (int j = 0; j < size; ++j) {
            board[i][j] = '~';
        }
    }
    return board;
}

void freeBoard(char **board, int size) {
    for (int i = 0; i < size; ++i) {
        free(board[i]);
    }
    free(board);
}

void initializeBoards(char **board1, char **board2, int size) {
    for (int i = 0; i < size; ++i) {
        for (int j = 0; j < size; ++j) {
            board1[i][j] = '~';
            board2[i][j] = '~';
        }
    }
}

void printHiddenBoard(char **board, int size) {
    printf("   ");
    for (int i = 0; i < size; ++i) {
        printf("%d ", i + 1);
    }
    printf("\n");

    for (int i = 0; i < size; ++i) {
        printf("%c |", 'A' + i);
        for (int j = 0; j < size; ++j) {
            if (board[i][j] == 'X' || board[i][j] == 'O') {
                printf("%c ", board[i][j]);
            } else {
                printf("~ ");
            }
        }
        printf("\n");
    }
}

void printPlayerInfo(Player *player, const char *playerName, int shipCount) {
    printf("%s's Info:\n", playerName);
    printf("Score: %d\n", player->score);
    for (int i = 0; i < shipCount; ++i) {
        printf("Ship %d - Size: %d, Hits: %d\n", i + 1, player->ships[i].size, player->ships[i].hits);
    }
}

void placeShip(Player *player, char **board, int size, int shipIndex) {
    int x, y;
    int direction;

    do {
        printf("Enter the starting coordinates for Ship %d (size %d): ", shipIndex + 1, player->ships[shipIndex].size);
        char row;
        scanf(" %c %d", &row, &y);
        x = row - 'A';

        printf("Downward (1) or Rightward (2)? ");
        scanf("%d", &direction);

        if (direction == 1) {
            if (x + player->ships[shipIndex].size > size) {
                printf("Invalid placement. Try again.\n");
            } else {
                for (int i = 0; i < player->ships[shipIndex].size; ++i) {
                    if (board[x + i][y - 1] != '~') {
                        printf("Invalid placement. Try again.\n");
                        break;
                    } else {
                        board[x + i][y - 1] = shipIndex + '1';
                    }
                }
            }
        } else if (direction == 2) {
            if (y + player->ships[shipIndex].size > size) {
                printf("Invalid placement. Try again.\n");
            } else {
                for (int i = 0; i < player->ships[shipIndex].size; ++i) {
                    if (board[x][y + i - 1] != '~') {
                        printf("Invalid placement. Try again.\n");
                        break;
                    } else {
                        board[x][y + i - 1] = shipIndex + '1';
                    }
                }
            }
        } else {
            printf("Invalid direction. Try again.\n");
        }
    } while (direction != 1 && direction != 2);
}

void placeShips(Player *player, char **board, int size, int shipCount) {
    for (int i = 0; i < shipCount; ++i) {
        player->ships[i].size = i + 3;
        player->ships[i].hits = 0;
        player->ships[i].isSunk = 0;

        printHiddenBoard(board, size);
        printPlayerInfo(player, "Player", shipCount);
        placeShip(player, board, size, i);
        CLEAR_SCREEN;  // Ekranı temizle
    }
}

int makeMove(Player *player, char **opponentBoard, int size, int shipCount) {
    int x, y;

    printf("Enter the coordinates to make a move: ");
    char row;
    scanf(" %c %d", &row, &y);
    x = row - 'A';

    if (x < 0 || x >= size || y < 1 || y > size) {
        printf("Invalid coordinates. Try again.\n");
        return makeMove(player, opponentBoard, size, shipCount);
    }

    if (opponentBoard[x][y - 1] == 'X' || opponentBoard[x][y - 1] == 'O') {
        printf("Already guessed that location. Try again.\n");
        return makeMove(player, opponentBoard, size, shipCount);
    }

    if (opponentBoard[x][y - 1] != '~') {
        int shipIndex = opponentBoard[x][y - 1] - '1';
        printf("Hit!\n");
        opponentBoard[x][y - 1] = 'X';
        player->ships[shipIndex].hits++;

        if (player->ships[shipIndex].hits == player->ships[shipIndex].size) {
            printf("You sank a ship!\n");
            player->ships[shipIndex].isSunk = 1;
            player->score += player->ships[shipIndex].size;

            int allSunk = 1;
            for (int i = 0; i < shipCount; ++i) {
                if (opponentBoard[x][y - 1] == i + '1' && !player->ships[i].isSunk) {
                    allSunk = 0;
                    break;
                }
            }

            if (allSunk) {
                printf("All opponent's ships are sunk! You win!\n");
                return 0;  // Oyunu bitir
            }
        }
    } else {
        printf("Miss!\n");
        opponentBoard[x][y - 1] = 'O';
    }

    return 1;  // Oyun devam ediyor
}
int main() {
    int boardSize;
    int shipCount;
    printf("This game was made by Yasin Karadavut with student number 152420211024 on 18.01.2024.\nEskisehir Osmangazi University Computer Enginnering Introduction to Programing Lab \n");
    printf("Enter the board size (between %d and %d): ", MIN_BOARD_SIZE, MAX_BOARD_SIZE);
    scanf("%d", &boardSize);
    
    if (boardSize < MIN_BOARD_SIZE || boardSize > MAX_BOARD_SIZE) {
        printf("Invalid board size. Exiting...\n");
        return 1;
    }

    printf("Enter the number of ships(max 5): ");
    scanf("%d", &shipCount);

    if (shipCount < 1 || shipCount > MAX_SHIP_TYPES) {
        printf("Invalid number of ships. Exiting...\n");
        return 1;
    }

    char **player1Board = createBoard(boardSize);
    char **player2Board = createBoard(boardSize);

    Player player1, player2;
    player1.score = 0;
    player2.score = 0;

    printf("Enter Player 1's name: ");
    scanf("%s", player1.name);

    printf("Enter Player 2's name: ");
    scanf("%s", player2.name);

    printf("%s, place your ships:\n", player1.name);
    placeShips(&player1, player1Board, boardSize, shipCount);

    printf("%s, place your ships:\n", player2.name);
    placeShips(&player2, player2Board, boardSize, shipCount);

    int currentPlayer = 1;

    while (1) {
        if (currentPlayer == 1) {
            CLEAR_SCREEN;
            printf("%s's Turn:\n", player1.name);
            printHiddenBoard(player2Board, boardSize);
            printPlayerInfo(&player1, player1.name, shipCount);
            if (!makeMove(&player1, player2Board, boardSize, shipCount)) {
                break;  // Oyun bitti
            }
        } else {
            CLEAR_SCREEN;
            printf("%s's Turn:\n", player2.name);
            printHiddenBoard(player1Board, boardSize);
            printPlayerInfo(&player2, player2.name, shipCount);
            if (!makeMove(&player2, player1Board, boardSize, shipCount)) {
                break;  // Oyun bitti
            }
        }

        currentPlayer = 3 - currentPlayer;  // Sıradaki oyuncu
    }

    printf("Final Scores:\n");
    printPlayerInfo(&player1, player1.name, shipCount);
    printPlayerInfo(&player2, player2.name, shipCount);

    freeBoard(player1Board, boardSize);
    freeBoard(player2Board, boardSize);

    return 0;
}